File: web/clustering.rs
use crate::types::{Detection, DEFAULT_CLUSTER_SIZE, IOU_THRESHOLD};
use dbscan::cluster;
use rayon::prelude::*;

pub fn cluster_boxes(detections: &mut Vec<Detection>) -> Vec<[i32; 4]> {
    if detections.is_empty() {
        return Vec::new();
    }

    let mut points = Vec::with_capacity(detections.len());
    points.par_extend(detections.par_iter().map(|det| {
        vec![
            det.bbox[0] as f64,
            det.bbox[1] as f64,
            det.bbox[2] as f64,
            det.bbox[3] as f64,
        ]
    }));

    let eps = calculate_median_size(detections).max(DEFAULT_CLUSTER_SIZE) * 0.5;
    let min_points = if detections.len() > 3 { 2 } else { 1 };

    let clusters = cluster(eps, min_points, &points);
    if !clusters
        .iter()
        .all(|c| matches!(c, dbscan::Classification::Noise))
    {
        return process_clusters(detections, clusters);
    }
    process_clusters(detections, cluster(eps * 1.5, min_points, &points))
}

fn calculate_median_size(detections: &[Detection]) -> f64 {
    let mut sizes: Vec<f64> = detections
        .iter()
        .map(|det| {
            let width = (det.bbox[2] - det.bbox[0]) as f64;
            let height = (det.bbox[3] - det.bbox[1]) as f64;
            (width * height).sqrt()
        })
        .collect();

    sizes.sort_by(safe_f64_cmp);
    sizes
        .get(sizes.len() / 2)
        .copied()
        .unwrap_or(DEFAULT_CLUSTER_SIZE)
}

fn process_clusters(
    detections: &[Detection],
    clusters: Vec<dbscan::Classification>,
) -> Vec<[i32; 4]> {
    let mut final_boxes = Vec::new();
    let mut cluster_map: std::collections::HashMap<usize, Vec<[i32; 4]>> =
        std::collections::HashMap::new();

    for (idx, classification) in clusters.iter().enumerate() {
        if idx >= detections.len() {
            continue;
        }

        match classification {
            dbscan::Classification::Core(cluster_id) | dbscan::Classification::Edge(cluster_id) => {
                cluster_map
                    .entry(*cluster_id)
                    .or_default()
                    .push(detections[idx].bbox);
            }
            dbscan::Classification::Noise => {
                handle_noise_point(&mut final_boxes, &mut cluster_map, detections[idx].bbox);
            }
        }
    }

    finalize_clusters(&mut final_boxes, cluster_map, detections)
}

fn handle_noise_point(
    final_boxes: &mut Vec<[i32; 4]>,
    cluster_map: &mut std::collections::HashMap<usize, Vec<[i32; 4]>>,
    bbox: [i32; 4],
) {
    let mut merged = false;
    for boxes in cluster_map.values_mut() {
        if boxes.iter().any(|existing_box| {
            let iou = calculate_iou(&bbox, existing_box);
            iou.is_finite() && iou > IOU_THRESHOLD
        }) {
            boxes.push(bbox);
            merged = true;
            break;
        }
    }

    if !merged {
        final_boxes.push(bbox);
    }
}

fn finalize_clusters(
    final_boxes: &mut Vec<[i32; 4]>,
    cluster_map: std::collections::HashMap<usize, Vec<[i32; 4]>>,
    detections: &[Detection],
) -> Vec<[i32; 4]> {
    for boxes in cluster_map.values() {
        if !boxes.is_empty() {
            final_boxes.push(merge_boxes(boxes));
        }
    }

    if final_boxes.is_empty() {
        let mut seen = std::collections::HashSet::new();
        final_boxes.extend(
            detections
                .iter()
                .filter(|det| seen.insert(det.bbox))
                .map(|det| det.bbox),
        );
    }

    final_boxes.to_vec()
}

pub fn safe_f64_cmp(a: &f64, b: &f64) -> std::cmp::Ordering {
    match (a.is_nan(), b.is_nan()) {
        (true, true) => std::cmp::Ordering::Equal,
        (true, false) => std::cmp::Ordering::Greater,
        (false, true) => std::cmp::Ordering::Less,
        (false, false) => a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal),
    }
}

#[inline(always)]
pub fn merge_boxes(boxes: &[[i32; 4]]) -> [i32; 4] {
    if boxes.is_empty() {
        return [0, 0, 0, 0];
    }

    let mut min_x = boxes[0][0];
    let mut min_y = boxes[0][1];
    let mut max_x = boxes[0][2];
    let mut max_y = boxes[0][3];

    for bbox in boxes.iter().skip(1) {
        min_x = min_x.min(bbox[0]);
        min_y = min_y.min(bbox[1]);
        max_x = max_x.max(bbox[2]);
        max_y = max_y.max(bbox[3]);
    }

    [min_x, min_y, max_x, max_y]
}

pub fn convert_to_corners(box_coords: &[i32; 4]) -> [i32; 4] {
    let [x_center, y_center, width, height] = *box_coords;
    [
        x_center - width / 2,  // x1
        y_center - height / 2, // y1
        x_center + width / 2,  // x2
        y_center + height / 2, // y2
    ]
}

pub fn calculate_iou(box1: &[i32; 4], box2: &[i32; 4]) -> f32 {
    let box1_corners = convert_to_corners(box1);
    let box2_corners = convert_to_corners(box2);

    let x_left = box1_corners[0].max(box2_corners[0]);
    let y_top = box1_corners[1].max(box2_corners[1]);
    let x_right = box1_corners[2].min(box2_corners[2]);
    let y_bottom = box1_corners[3].min(box2_corners[3]);

    if x_right <= x_left || y_bottom <= y_top {
        return 0.0;
    }

    let intersection_area = ((x_right - x_left) * (y_bottom - y_top)) as f32;
    let box1_area = (box1[2] * box1[3]) as f32;
    let box2_area = (box2[2] * box2[3]) as f32;
    let union_area = box1_area + box2_area - intersection_area;

    intersection_area / union_area
}

File: web/clustering.go
package main

import (
	"math"
	"sort"
)

const (
	DefaultClusterSize = 50.0
	IouThreshold       = 0.45
)

func clusterBoxes(detections []Detection) [][4]int32 {
	if len(detections) == 0 {
		return nil
	}

	medianSize := calculateMedianSize(detections)
	eps := math.Max(medianSize, DefaultClusterSize) * 0.5
	minPoints := 1
	if len(detections) > 3 {
		minPoints = 2
	}

	points := make([][]float64, len(detections))
	for i, det := range detections {
		points[i] = []float64{
			float64(det.BBox[0]),
			float64(det.BBox[1]),
			float64(det.BBox[2]),
			float64(det.BBox[3]),
		}
	}

	clusters := dbscan(points, eps, minPoints)
	return processClusters(detections, clusters)
}

func calculateMedianSize(detections []Detection) float64 {
	sizes := make([]float64, len(detections))
	for i, det := range detections {
		width := float64(det.BBox[2] - det.BBox[0])
		height := float64(det.BBox[3] - det.BBox[1])
		sizes[i] = math.Sqrt(width * height)
	}

	sort.Float64s(sizes)
	if len(sizes) == 0 {
		return DefaultClusterSize
	}
	return sizes[len(sizes)/2]
}

func processClusters(detections []Detection, clusters []int) [][4]int32 {
	clusterMap := make(map[int][][4]int32)
	var finalBoxes [][4]int32

	// Group boxes by cluster
	for i, cluster := range clusters {
		if cluster == -1 {
			// Handle noise points
			bbox := detections[i].BBox
			merged := false
			for _, boxes := range clusterMap {
				for _, existing := range boxes {
					if calculateIOU(bbox, existing) > IouThreshold {
						boxes = append(boxes, bbox)
						merged = true
						break
					}
				}
				if merged {
					break
				}
			}
			if !merged {
				finalBoxes = append(finalBoxes, bbox)
			}
		} else {
			clusterMap[cluster] = append(clusterMap[cluster], detections[i].BBox)
		}
	}

	// Merge boxes in each cluster
	for _, boxes := range clusterMap {
		if len(boxes) > 0 {
			finalBoxes = append(finalBoxes, mergeBoxes(boxes))
		}
	}

	return finalBoxes
}

func calculateIOU(box1, box2 [4]int32) float64 {
	x1 := math.Max(float64(box1[0]), float64(box2[0]))
	y1 := math.Max(float64(box1[1]), float64(box2[1]))
	x2 := math.Min(float64(box1[2]), float64(box2[2]))
	y2 := math.Min(float64(box1[3]), float64(box2[3]))

	if x2 <= x1 || y2 <= y1 {
		return 0.0
	}

	intersection := (x2 - x1) * (y2 - y1)
	area1 := float64((box1[2] - box1[0]) * (box1[3] - box1[1]))
	area2 := float64((box2[2] - box2[0]) * (box2[3] - box2[1]))
	union := area1 + area2 - intersection

	return intersection / union
}

func mergeBoxes(boxes [][4]int32) [4]int32 {
	if len(boxes) == 0 {
		return [4]int32{0, 0, 0, 0}
	}

	result := boxes[0]
	for _, box := range boxes[1:] {
		result[0] = min32(result[0], box[0])
		result[1] = min32(result[1], box[1])
		result[2] = max32(result[2], box[2])
		result[3] = max32(result[3], box[3])
	}

	return result
}

func min32(a, b int32) int32 {
	if a < b {
		return a
	}
	return b
}

func max32(a, b int32) int32 {
	if a > b {
		return a
	}
	return b
}

func dbscan(points [][]float64, eps float64, minPoints int) []int {
	n := len(points)
	clusters := make([]int, n)
	for i := range clusters {
		clusters[i] = -1 // Initialize all points as noise
	}

	currentCluster := 0
	for i := 0; i < n; i++ {
		if clusters[i] != -1 {
			continue
		}

		neighbors := getNeighbors(points, i, eps)
		if len(neighbors) < minPoints {
			continue
		}

		clusters[i] = currentCluster
		expandCluster(points, clusters, neighbors, currentCluster, eps, minPoints)
		currentCluster++
	}

	return clusters
}

func getNeighbors(points [][]float64, pointIdx int, eps float64) []int {
	var neighbors []int
	for i := range points {
		if distance(points[pointIdx], points[i]) <= eps {
			neighbors = append(neighbors, i)
		}
	}
	return neighbors
}

func expandCluster(points [][]float64, clusters []int, neighbors []int, cluster int, eps float64, minPoints int) {
	for i := 0; i < len(neighbors); i++ {
		pointIdx := neighbors[i]
		if clusters[pointIdx] == -1 {
			clusters[pointIdx] = cluster
			newNeighbors := getNeighbors(points, pointIdx, eps)
			if len(newNeighbors) >= minPoints {
				neighbors = append(neighbors, newNeighbors...)
			}
		}
	}
}

func distance(p1, p2 []float64) float64 {
	sum := 0.0
	for i := range p1 {
		diff := p1[i] - p2[i]
		sum += diff * diff
	}
	return math.Sqrt(sum)
}
